lib/internal/wasi.js

```diff
@@ -0,0 +1,1199 @@
+  ENOTCONN: WASI_ENOTCONN,
+  ENOTDIR: WASI_ENOTDIR,
+  ENOTEMPTY: WASI_ENOTEMPTY,
+  ENOTRECOVERABLE: WASI_ENOTRECOVERABLE,
+  ENOTSOCK: WASI_ENOTSOCK,
+  ENOTTY: WASI_ENOTTY,
+  // ENOTUNIQ: WASI_ENOTUNIQ,
+  ENXIO: WASI_ENXIO,
+  EOVERFLOW: WASI_EOVERFLOW,
+  EOWNERDEAD: WASI_EOWNERDEAD,
+  EPERM: WASI_EPERM,
+  // EPFNOSUPPORT: WASI_EPFNOSUPPORT,
+  EPIPE: WASI_EPIPE,
+  EPROTO: WASI_EPROTO,
+  EPROTONOSUPPORT: WASI_EPROTONOSUPPORT,
+  EPROTOTYPE: WASI_EPROTOTYPE,
+  ERANGE: WASI_ERANGE,
+  // EREMCHG: WASI_EREMCHG,
+  // EREMOTE: WASI_EREMOTE,
+  // EREMOTEIO: WASI_EREMOTEIO,
+  // ERESTART: WASI_ERESTART,
+  // ERFKILL: WASI_ERFKILL,
+  EROFS: WASI_EROFS,
+  // ESHUTDOWN: WASI_ESHUTDOWN,
+  ESPIPE: WASI_ESPIPE,
+  // ESOCKTNOSUPPORT: WASI_ESOCKTNOSUPPORT,
+  ESRCH: WASI_ESRCH,
+  ESTALE: WASI_ESTALE,
+  // ESTRPIPE: WASI_ESTRPIPE,
+  // ETIME: WASI_ETIME,
+  ETIMEDOUT: WASI_ETIMEDOUT,
+  // ETOOMANYREFS: WASI_ETOOMANYREFS,
+  ETXTBSY: WASI_ETXTBSY,
+  // EUCLEAN: WASI_EUCLEAN,
+  // EUNATCH: WASI_EUNATCH,
+  // EUSERS: WASI_EUSERS,
+  EXDEV: WASI_EXDEV,
+  // EXFULL: WASI_EXFULL,
+};
+
+const SIGNAL_MAP = {
+  __proto__: null,
+  [WASI_SIGHUP]: 'SIGHUP',
+  [WASI_SIGINT]: 'SIGINT',
+  [WASI_SIGQUIT]: 'SIGQUIT',
+  [WASI_SIGILL]: 'SIGILL',
+  [WASI_SIGTRAP]: 'SIGTRAP',
+  [WASI_SIGABRT]: 'SIGABRT',
+  // [WASI_SIGIOT]: 'SIGIOT',
+  [WASI_SIGBUS]: 'SIGBUS',
+  [WASI_SIGFPE]: 'SIGFPE',
+  [WASI_SIGKILL]: 'SIGKILL',
+  [WASI_SIGUSR1]: 'SIGUSR1',
+  [WASI_SIGSEGV]: 'SIGSEGV',
+  [WASI_SIGUSR2]: 'SIGUSR2',
+  [WASI_SIGPIPE]: 'SIGPIPE',
+  [WASI_SIGALRM]: 'SIGALRM',
+  [WASI_SIGTERM]: 'SIGTERM',
+  [WASI_SIGCHLD]: 'SIGCHLD',
+  [WASI_SIGCONT]: 'SIGCONT',
+  [WASI_SIGSTOP]: 'SIGSTOP',
+  [WASI_SIGTSTP]: 'SIGTSTP',
+  [WASI_SIGTTIN]: 'SIGTTIN',
+  [WASI_SIGTTOU]: 'SIGTTOU',
+  [WASI_SIGURG]: 'SIGURG',
+  [WASI_SIGXCPU]: 'SIGXCPU',
+  [WASI_SIGXFSZ]: 'SIGXFSZ',
+  [WASI_SIGVTALRM]: 'SIGVTALRM',
+  // [WASI_SIGPROF]: 'SIGPROF',
+  // [WASI_SIGWINCH]: 'SIGWINCH',
+  // [WASI_SIGIO]: 'SIGIO',
+  // [WASI_SIGINFO]: 'SIGINFO',
+  // [WASI_SIGSYS]: 'SIGSYS',
+};
+
+const convertOpenFlags = (flags) => {
+  let out = 0;
+  if (flags & WASI_O_CREAT) {
+    out |= fs.constants.O_CREAT;
+  }
+  if (flags & WASI_O_DIRECTORY) {
+    out |= fs.constants.O_DIRECTORY;
+  }
+  if (flags & WASI_O_EXCL) {
+    out |= fs.constants.O_EXCL;
+  }
+  if (flags & WASI_O_TRUNC) {
+    out |= fs.constants.O_TRUNC;
+  }
+  return out;
+};
+
+const now = (clockId) => {
+  switch (clockId) {
+    case WASI_CLOCK_MONOTONIC:
+    case WASI_CLOCK_REALTIME:
+      return process.hrtime.bigint();
+    case WASI_CLOCK_PROCESS_CPUTIME_ID:
+    case WASI_CLOCK_THREAD_CPUTIME_ID:
+      return process.hrtime.bigint(CPUTIME_START);
+    default:
+      return null;
+  }
+};
+
+const msToNs = (ms) => {
+  const decimal = BigInt(
+    Math.trunc(Number.parseFloat((ms % 1).toFixed(3), 10) * 1000),
+  );
+  const ns = BigInt(Math.trunc(ms)) * 1000n;
+  return ns + decimal;
+};
```

Comment:
What about speedup this a little?
```ts
const msToNs = ms => {
  const msInt = Math.trunc(ms)
  const decimal = BigInt(Math.round((ms - msInt) * 1000));
  const ns = BigInt(msInt) * 1000n;
  return ns + decimal;
};
```

Reply:
would you mind making a pr to https://github.com/devsnek/node-wasi

Reply:
Sure!